File: py/lc/1248_count_nice_arrays.py
"""
https://leetcode.com/problems/count-number-of-nice-subarrays/description/?envType=company&envId=roblox&favoriteSlug=roblox-all

Given an array of integers nums and an integer k. A continuous subarray is called nice if there are k odd numbers on it.
Return the number of nice sub-arrays.

Example 1:
Input: nums = [1,1,2,1,1], k = 3
Output: 2
Explanation: The only sub-arrays with 3 odd numbers are [1,1,2,1] and [1,2,1,1].

Example 2:
Input: nums = [2,4,6], k = 1
Output: 0
Explanation: There are no odd numbers in the array.

Example 3:
Input: nums = [2,2,2,1,2,2,1,2,2,2], k = 2
Output: 16

Constraints:
1 <= nums.length <= 50000
1 <= nums[i] <= 10^5
1 <= k <= nums.length
"""

from typing import List


class Solution:
    def number_of_subarrays(self, nums: List[int], k: int) -> int:
        lp = 0
        even, odd = 0, 0
        result = 0

        for rp in range(len(nums)):
            if nums[rp] % 2 == 1:
                odd += 1
                even = 0

            while odd == k:
                if nums[lp] % 2 == 1:
                    odd -= 1
                lp += 1
                even += 1

            result += even

        return result


def test_one():
    solution = Solution()

    assert solution.number_of_subarrays([1, 1, 2, 1, 1], 3) == 2


def test_two():
    solution = Solution()

    assert solution.number_of_subarrays([2, 4, 6], 1) == 0


def test_three():
    solution = Solution()

    assert solution.number_of_subarrays([2, 2, 2, 1, 2, 2, 1, 2, 2, 2], 2) == 16

File: rs/src/lc/1248_count_nice_arrays.rs
pub fn number_of_subarrays(nums: Vec<i32>, k: i32) -> i32 {
    let mut lp = 0;
    let mut odd_count = 0;
    let mut even_count = 0;
    let mut result = 0;

    for rp in 0..nums.len() {
        if nums[rp] % 2 != 0 {
            odd_count += 1;
            even_count = 0;
        }

        while odd_count == k {
            if nums[lp] % 2 != 0 {
                odd_count -= 1;
            }
            lp += 1;
            even_count += 1;
        }

        result += even_count;
    }

    result
}

#[test]
fn test_one() {
    let result = number_of_subarrays(vec![1, 1, 2, 1, 1], 3);

    assert_eq!(result, 2);
}

#[test]
fn test_two() {
    let result = number_of_subarrays(vec![2, 4, 6], 1);

    assert_eq!(result, 0);
}

#[test]
fn test_three() {
    let result = number_of_subarrays(vec![2, 2, 2, 1, 2, 2, 1, 2, 2, 2], 2);

    assert_eq!(result, 16)
}
